lib/auth/auth.go

```diff
@@ -1529,6 +1528,21 @@ func (a *Server) runPeriodicOperations() {
 									}
 								}
 
+								// If the hostname has been replaced by a sanitized version, revert it back to the original
+								// if the original is valid under the new rules.
+								if oldHostname, ok := srv.GetLabel(replacedHostnameLabel); ok && validServerHostname(oldHostname) {
+									switch s := srv.(type) {
+									case *types.ServerV2:
+										s.Spec.Hostname = oldHostname
+										delete(s.Metadata.Labels, replacedHostnameLabel)
+									default:
+										return false, trace.BadParameter("invalid server provided")
```

Comment:
The previous sanitization doesn't stop the iteration on error, why should this one?